jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/ResourceFactoryInternals.java

```diff
@@ -29,103 +33,169 @@
-        {
-            return FileSystemPool.INSTANCE.mount(uri);
-        }
-        catch (IOException ioe)
+        @Override
+        protected void onMounted(FileSystemPool.Mount mount, URI uri)
         {
-            throw new IllegalArgumentException(ioe);
+            // Since this ROOT ResourceFactory and has no lifecycle that can clean up
+            // the mount, we shall report this mount as a leak
+            if (LOG.isDebugEnabled())
+                LOG.warn("Leaked {} for {}", mount, uri, new Throwable());
+            else
+                LOG.warn("Leaked {} for {}", mount, uri);
         }
-    }
+    };
 
     static class Closeable implements ResourceFactory.Closeable
     {
-        private final List<FileSystemPool.Mount> _mounts = new CopyOnWriteArrayList<>();
+        private final CompositeResourceFactory _compositeResourceFactory = new CompositeResourceFactory();
 
         @Override
         public Resource newResource(URI uri)
         {
-            FileSystemPool.Mount mount = mountIfNeeded(uri);
-            if (mount != null)
-                _mounts.add(mount);
-            return Resource.create(uri);
+            return _compositeResourceFactory.newResource(uri);
         }
 
         @Override
         public void close()
         {
-            for (FileSystemPool.Mount mount : _mounts)
+            for (FileSystemPool.Mount mount : _compositeResourceFactory.getMounts())
                 IO.close(mount);
-            _mounts.clear();
+            _compositeResourceFactory.clearMounts();
         }
     }
 
     static class LifeCycle extends AbstractLifeCycle implements ResourceFactory.LifeCycle
     {
-        private final List<FileSystemPool.Mount> _mounts = new CopyOnWriteArrayList<>();
+        private final CompositeResourceFactory _compositeResourceFactory = new CompositeResourceFactory();
 
         @Override
         public Resource newResource(URI uri)
         {
-            FileSystemPool.Mount mount = mountIfNeeded(uri);
-            if (mount != null)
-                _mounts.add(mount);
-            return Resource.create(uri);
+            return _compositeResourceFactory.newResource(uri);
         }
 
         @Override
         protected void doStop() throws Exception
         {
-            for (FileSystemPool.Mount mount : _mounts)
+            for (FileSystemPool.Mount mount : _compositeResourceFactory.getMounts())
                 IO.close(mount);
-            _mounts.clear();
+            _compositeResourceFactory.clearMounts();
             super.doStop();
         }
 
         @Override
         public void dump(Appendable out, String indent) throws IOException
         {
-            Dumpable.dumpObjects(out, indent, this, new DumpableCollection("mounts", _mounts));
+            Dumpable.dumpObjects(out, indent, this, new DumpableCollection("mounts", _compositeResourceFactory.getMounts()));
+        }
+    }
+
+    static class CompositeResourceFactory implements ResourceFactory
```

Comment:
I know this is not a public name.... but `CompositeResourceFactory` is not really correct.
How about `DelegatingResourceFactory` as it delegates to a another factory.
Or maybe `BySchemeResourceFactory` as it chooses another factory by scheme.
Or to be absolutely precise `DelegatingBySchemeResourceFactory`.

Anything to avoid the contentious and misunderstood "Composite"!

Reply:
how about `UriResourceFactory`?